File: src/utils/git-policy.ts
/**
 * Git Policy — Freddie Mac Version Control Standards
 *
 * Encapsulates enterprise Git governance rules:
 * - Branch naming conventions (GitFlow)
 * - Commit message format enforcement
 * - Pre-push validation checklist
 */

/** Valid branch prefixes per GitFlow convention */
const BRANCH_PREFIXES = ["feature/", "release/", "hotfix/", "bugfix/"] as const;
const PROTECTED_BRANCHES = ["main", "master", "develop"] as const;

/** Conventional commit prefixes */
const COMMIT_PREFIXES = [
  "feat:",
  "fix:",
  "hotfix:",
  "chore:",
  "docs:",
  "refactor:",
  "test:",
  "ci:",
  "perf:",
] as const;

export interface PolicyViolation {
  rule: string;
  message: string;
  severity: "error" | "warning";
}

/**
 * Validates a branch name against Freddie Mac GitFlow standards.
 *
 * Valid patterns:
 *   feature/auth-101
 *   release/1.2.0
 *   hotfix/sec-fix-001
 *   bugfix/login-error
 *   main, develop (protected)
 */
export function validateBranchName(branchName: string): PolicyViolation[] {
  const violations: PolicyViolation[] = [];
  const normalized = branchName.trim().toLowerCase();

  // Protected branches are always valid
  if (
    PROTECTED_BRANCHES.includes(
      normalized as (typeof PROTECTED_BRANCHES)[number],
    )
  ) {
    return violations;
  }

  // Must start with a valid prefix
  const hasValidPrefix = BRANCH_PREFIXES.some((prefix) =>
    normalized.startsWith(prefix),
  );
  if (!hasValidPrefix) {
    violations.push({
      rule: "GIT-001",
      message: `Branch name '${branchName}' does not follow GitFlow convention. Must start with: ${BRANCH_PREFIXES.join(", ")}`,
      severity: "error",
    });
  }

  // No uppercase allowed
  if (branchName !== branchName.toLowerCase()) {
    violations.push({
      rule: "GIT-002",
      message: `Branch name '${branchName}' contains uppercase characters. Use lowercase with hyphens.`,
      severity: "error",
    });
  }

  // No spaces or special characters
  if (/[^a-z0-9\-\/._]/.test(normalized)) {
    violations.push({
      rule: "GIT-003",
      message: `Branch name '${branchName}' contains invalid characters. Use only alphanumeric, hyphens, dots, and forward slashes.`,
      severity: "error",
    });
  }

  // Must have a descriptor after the prefix
  const afterPrefix = BRANCH_PREFIXES.reduce((result, prefix) => {
    return normalized.startsWith(prefix)
      ? normalized.slice(prefix.length)
      : result;
  }, normalized);

  if (hasValidPrefix && afterPrefix.length === 0) {
    violations.push({
      rule: "GIT-004",
      message: `Branch name '${branchName}' is missing a descriptor after the prefix (e.g., feature/auth-101).`,
      severity: "error",
    });
  }

  return violations;
}

/**
 * Validates a commit message against conventional commit format.
 *
 * Valid patterns:
 *   feat: Add login endpoint
 *   fix: Resolve null pointer in auth
 *   hotfix: Patch critical vulnerability
 *   chore: Update dependencies
 */
export function validateCommitMessage(message: string): PolicyViolation[] {
  const violations: PolicyViolation[] = [];
  const trimmed = message.trim();

  if (trimmed.length === 0) {
    violations.push({
      rule: "GIT-010",
      message: "Commit message cannot be empty.",
      severity: "error",
    });
    return violations;
  }

  // Must start with a conventional prefix
  const hasValidPrefix = COMMIT_PREFIXES.some((prefix) =>
    trimmed.toLowerCase().startsWith(prefix),
  );
  if (!hasValidPrefix) {
    violations.push({
      rule: "GIT-011",
      message: `Commit message does not follow conventional format. Must start with: ${COMMIT_PREFIXES.join(", ")}`,
      severity: "error",
    });
  }

  // Subject line should not exceed 72 characters
  const firstLine = trimmed.split("\n")[0];
  if (firstLine.length > 72) {
    violations.push({
      rule: "GIT-012",
      message: `Commit subject line exceeds 72 characters (${firstLine.length}). Keep it concise.`,
      severity: "warning",
    });
  }

  // Subject should not end with a period
  if (firstLine.endsWith(".")) {
    violations.push({
      rule: "GIT-013",
      message: "Commit subject line should not end with a period.",
      severity: "warning",
    });
  }

  return violations;
}

/**
 * Returns the GitFlow branching strategy documentation.
 */
export function getGitFlowGuide(): string[] {
  return [
    "Freddie Mac GitFlow Branching Strategy:",
    "",
    "  main          Production-ready code. Protected, no direct commits.",
    "  develop       Integration branch. Features merge here first.",
    "  feature/*     New features branched from develop.",
    "  release/*     Release preparation branched from develop.",
    "  hotfix/*      Emergency patches branched from main.",
    "  bugfix/*      Bug fixes branched from develop.",
    "",
    "Workflow:",
    "  1. Create feature branch:  feature/<ticket-id>",
    "  2. Develop and commit with conventional messages",
    "  3. Merge to develop with --no-ff (preserve history)",
    "  4. Cut release branch: release/<version>",
    "  5. Test and stabilize on release branch",
    "  6. Merge release to main AND develop",
    "  7. Tag the release on main",
    "",
    "Rollback Strategy:",
    "  - Use 'git revert' (non-destructive, preserves audit trail)",
    "  - Never use 'git reset --hard' on shared branches",
    "  - All reversals create a new commit for traceability",
  ];
}

/**
 * Generates the conventional commit message for a ticket.
 */
export function buildCommitMessage(
  ticketId: string,
  title: string,
  type: "feat" | "fix" | "hotfix" | "chore" = "feat",
): string {
  return `${type}: [${ticketId.toUpperCase()}] ${title}`;
}

/**
 * Returns the expected branch name for a given ticket and type.
 */
export function buildBranchName(
  ticketId: string,
  type: "feature" | "hotfix" | "bugfix" = "feature",
): string {
  return `${type}/${ticketId.toLowerCase()}`;
}

File: src/utils/git_policy.py
"""Git Policy — GitFlow branching & conventional commits. Replaces git-policy.ts."""
from __future__ import annotations
import re
from ..core.types import PolicyViolation

BRANCH_PREFIXES = ("feature/", "release/", "hotfix/", "bugfix/")
PROTECTED_BRANCHES = ("main", "master", "develop")
COMMIT_PREFIXES = ("feat:", "fix:", "hotfix:", "chore:", "docs:", "refactor:", "test:", "ci:", "perf:")

def validate_branch_name(name: str) -> list[PolicyViolation]:
    v, n = [], name.strip().lower()
    if n in PROTECTED_BRANCHES: return v
    if not any(n.startswith(p) for p in BRANCH_PREFIXES):
        v.append(PolicyViolation("GIT-001", f"Branch '{name}' must start with: {', '.join(BRANCH_PREFIXES)}", "error"))
    if name != name.lower():
        v.append(PolicyViolation("GIT-002", f"Branch '{name}' has uppercase chars. Use lowercase.", "error"))
    if re.search(r"[^a-z0-9\-/._]", n):
        v.append(PolicyViolation("GIT-003", f"Branch '{name}' has invalid characters.", "error"))
    after = n
    for p in BRANCH_PREFIXES:
        if n.startswith(p): after = n[len(p):]; break
    if any(n.startswith(p) for p in BRANCH_PREFIXES) and not after:
        v.append(PolicyViolation("GIT-004", f"Branch '{name}' missing descriptor after prefix.", "error"))
    return v

def validate_commit_message(msg: str) -> list[PolicyViolation]:
    v, t = [], msg.strip()
    if not t: return [PolicyViolation("GIT-010", "Commit message cannot be empty.", "error")]
    if not any(t.lower().startswith(p) for p in COMMIT_PREFIXES):
        v.append(PolicyViolation("GIT-011", f"Commit must start with: {', '.join(COMMIT_PREFIXES)}", "error"))
    first = t.split("\n")[0]
    if len(first) > 72: v.append(PolicyViolation("GIT-012", f"Subject line exceeds 72 chars ({len(first)}).", "warning"))
    if first.endswith("."): v.append(PolicyViolation("GIT-013", "Subject should not end with period.", "warning"))
    return v

def get_gitflow_guide() -> list[str]:
    return ["GitFlow Branching Strategy:", "", "  main       Production-ready. Protected.", "  develop    Integration branch.",
        "  feature/*  New features from develop.", "  release/*  Release prep from develop.", "  hotfix/*   Emergency patches from main.",
        "  bugfix/*   Bug fixes from develop.", "", "Workflow:", "  1. Create feature branch: feature/<ticket-id>",
        "  2. Develop with conventional commits", "  3. Merge to develop with --no-ff", "  4. Cut release: release/<version>",
        "  5. Merge release to main AND develop", "  6. Tag the release on main", "",
        "Rollback: Use 'git revert' (non-destructive, audit trail preserved)."]

def build_commit_message(ticket_id: str, title: str, commit_type: str = "feat") -> str:
    return f"{commit_type}: [{ticket_id.upper()}] {title}"

def build_branch_name(ticket_id: str, branch_type: str = "feature") -> str:
    return f"{branch_type}/{ticket_id.lower()}"
